File: ConfAndPlug/cdmwinit/java/src/cdmw/cdmwinit/CdmwInitConfiguration.java
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES 
 * and SELEX-SI. 
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003. 
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under 
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your 
 * option) any later version. 
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public 
 * License for more details. 
 * 
 * You should have received a copy of the GNU Library General 
 * Public License along with CARDAMOM; see the file COPYING. If not, write to 
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


package cdmw.cdmwinit;

import cdmw.common.InternalErrorException;
import cdmw.common.BadParameterException;
import cdmw.common.Assert;

import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

import org.xml.sax.ErrorHandler;
import org.xml.sax.SAXException;
import org.xml.sax.SAXParseException;
import org.apache.xerces.parsers.DOMParser;

/**
 * This class serves as a facility for accessing to configuration information
 * that is provided in an XML file and that is needed by the CDMW init.
 *
 */
public class CdmwInitConfiguration {

    /**
     * Default XML parser feature
     */
    private static boolean setNameSpaces    = false;

    /**
     * Default XML parser feature
     */
    private static boolean setSchemaSupport = true;

    /**
     * The DOM parser
     */
    protected DOMParser parser;

    /**
     * The DOM error handler
     */
    protected ErrorHandler errorReporter;

    /**
     * The parent node for all services. This node
     * is the starting point for the getServiceAttribute method.
     */
    protected Node servicesNode = null;

    /**
     * The parent node for all factories. This node
     * is the starting point for the getFactoryAttribute method.
     */
    protected Node lifeCycleServiceNode = null;

    /**
     * Constructs a CdmwInitConfiguration
     * by initializing the underlying XML parser.
     *
     * @exception ConfigurationException Configuration exception
     * @exception InternalErrorException Error during XML parsing
     */
    protected CdmwInitConfiguration()
        throws ConfigurationException, InternalErrorException {

        this(false);

    }

    /**
     * Constructs a CdmwInitConfiguration
     * by initializing the underlying XML parser.
     *
     * @exception ConfigurationException Configuration exception
     * @exception InternalErrorException Error during XML parsing
     */
    protected CdmwInitConfiguration(boolean doValidate)
        throws ConfigurationException, InternalErrorException {

        try {
            //
            //  Create our parser, then attach an error handler to the parser.
            //  The parser will call back to methods of the ErrorHandler if it
            //  discovers errors during the course of parsing the XML document.
            //
            this.parser = new DOMParser();

            if (doValidate) { // TODO: Check exceptions to be thrown...
                parser.setFeature(
                    "http://xml.org/sax/features/namespaces", setNameSpaces );
                parser.setFeature(
                    "http://apache.org/xml/features/validation/schema", setSchemaSupport );
            }

            DOMTreeErrorReporter errorHandler = new DOMTreeErrorReporter();
            parser.setErrorHandler(errorHandler);

        } catch(SAXException se) {
            throw new InternalErrorException();
        }

    }

    /**
     * Sets up the init configuration with the
     * specified XML configuration file
     *
     * @return A CdmwInitConfiguration object.
     *
     * @param XMLConfigurationFile the XML configuration file
     * @param doValidate XML file validation option
     *
     * @exception ConfigurationException Configuration exception
     * @exception InternalErrorException Error during XML parsing
     */
    public static CdmwInitConfiguration
        readConfiguration(String XMLConfigurationFile,boolean doValidate)
        throws ConfigurationException, InternalErrorException {

        CdmwInitConfiguration conf = new CdmwInitConfiguration(doValidate);
        conf.readConfigurationFile(XMLConfigurationFile);
        return conf;

    }


    /**
     * Gets the value of the specified service's attribute
     *
     * @param service   The service that owns the attribute
     * @param attribute The attribute whose value is queried
     *
     * @return The value of the specified attribute
     *
     * @exception BadParameterException Bad parameter value
     * @exception XMLErrorException Error during XML parsing
     */
    public String getServiceAttributeValue(String service, String attribute)
        throws BadParameterException,XMLErrorException {

        boolean serviceFound = false;
        boolean attributeFound = false;

        String attributeValue = null;

        Assert.check(servicesNode != null);

        // search the specified service among the children nodes
        Node serviceNode = findNode(servicesNode, service);

        if (serviceNode != null) {
            serviceFound = true;

            // search the specified attribute
            NamedNodeMap attributes = serviceNode.getAttributes();
            Node attribName = attributes.getNamedItem(attribute);

            if (attribName!=null) {
                // attribute found
                attributeFound = true;
                attributeValue = attribName.getNodeValue();
            }

        }

        if (!serviceFound) {
            throw new BadParameterException("service not found");
        }

        if (!attributeFound) {
            throw new BadParameterException("attribute not found");
        }

        return attributeValue;

    }


    /**
     * Gets the value of the specified factory's attribute corresponding to :
     * - the specified factory type
     * - specified service's attribute
     *
     * @param framework         The factory type
     * @param frameworkInstance The index of the factory instance
     * @param attribute           The attribute
     *
     * @return The value of the specified attribute
     *
     * @exception BadParameterException Bad parameter value
     * @exception XMLErrorException Error during XML parsing
     */
    public String getFrameworkAttributeValue(String framework,
            int frameworkInstance, String attribute)
        throws BadParameterException,XMLErrorException {

        boolean frameworkFound = false;
        boolean frameworkInstanceFound = false;
        boolean attributeFound = false;

        String attributeValue = null;

        Assert.check(lifeCycleServiceNode!=null);

        // search the specified service among the children nodes
        Node frameworkNode = findNode(lifeCycleServiceNode, framework);

        if (frameworkNode!=null) {

            frameworkFound = true;

            // search the specified factory instance
            Node instance
                = getActualChildNode(frameworkNode, frameworkInstance);

            if (instance != null) {

                frameworkInstanceFound = true;

                // search the specified attribute
                NamedNodeMap attributes = instance.getAttributes();
                Node attribName = attributes.getNamedItem(attribute);

                if (attribName!=null) {
                    // attribute found
                    attributeFound = true;
                    attributeValue = attribName.getNodeValue();
                }

            }

        }

        if (!frameworkFound) {
            throw new BadParameterException(framework + " framework not found");
        }

        if (!frameworkInstanceFound) {
            throw new BadParameterException("frameworkInstance invalid index");
        }

        if (!attributeFound) {
            throw new BadParameterException(attribute + " attribute not found");
        }

        return attributeValue;

    }

    /**
     * Gets the property (the unique attribute of the specified factory)
     * value of the specified factory's attribute corresponding to :
     * - the specified factory type
     * - specified service's attribute
     *
     * @param framework     The factory type
     * @param frameworkInstance The index of the factory instance
     *
     * @return The values of the specified attribute, the property
     * and the attribute in a String table
     *
     * @exception BadParameterException Bad parameter value
     * @exception XMLErrorException Error during XML parsing
     */
    public String[] getFrameworkMultithreadAttribute(String framework,
                                               int frameworkInstance)
        throws BadParameterException,XMLErrorException {

        boolean frameworkFound = false;
        boolean frameworkInstanceFound = false;
        boolean multithreadFound = false;
        boolean policyFound = false;

        String threadingPolicy = null;
        String threadingAttribute = null;
        String attributeValue = null;

        Assert.check(lifeCycleServiceNode!=null);

        // search the specified service among the children nodes
        Node frameworkNode = findNode(lifeCycleServiceNode, framework);

        if (frameworkNode!=null) {

            frameworkFound = true;

            // search the specified factory instance
            Node instance
                = getActualChildNode(frameworkNode, frameworkInstance);

            if (instance!=null) {

                frameworkInstanceFound = true;

                Node multithreadNode = findNode(instance, "multithread");

                if (multithreadNode!=null) {

                    multithreadFound = true;

                    Node policyNode
                        = getActualChildNode(multithreadNode, 0);
                    if (policyNode!=null) {
                        policyFound = true;
                        threadingPolicy = policyNode.getNodeName();

                        // search the unique attribute
                        NamedNodeMap attributes
                            = policyNode.getAttributes();
                        Node attribute = attributes.item(0);

                        if (attribute!=null) {
                            threadingAttribute = attribute.getNodeName();
                            attributeValue = attribute.getNodeValue();
                        } else {
                            // the threading attribute is irrelevant
                            threadingAttribute = "";
                            attributeValue = "";
                        }

                    }

                }

            }

        }

        if (!frameworkFound) {
            throw new BadParameterException(framework + " framework not found");
        }

        if (!frameworkInstanceFound) {
            throw new BadParameterException("frameworkInstance invalid index");
        }

        if (!multithreadFound) {
            throw new BadParameterException(framework + " not multithreaded framework");
        }

        if (!policyFound) {
            throw new BadParameterException("Threading policy: policy not found");
        }

        String[] values = {
            threadingPolicy,
            threadingAttribute,
            attributeValue
        };
        return values;

    }

    /**
     * Sets up the init configuration with the
     * specified XML configuration file
     *
     * @param XMLConfigurationFile the XML configuration file
     *
     * @exception XMLErrorException Error during XML parsing
     */
    protected void readConfigurationFile(String XMLConfigurationFile)
        throws XMLErrorException {

        try {
            parser.parse(XMLConfigurationFile);

            Document document = parser.getDocument();

            Element rootElement = document.getDocumentElement();

            servicesNode = findNode(rootElement, "services");

            lifeCycleServiceNode = findNode(servicesNode, "lifecycle");

        } catch (SAXException saxe) {
            saxe.printStackTrace();
            throw new XMLErrorException(saxe.getMessage());
        } catch (java.io.IOException ioe) {
            ioe.printStackTrace();
            throw new XMLErrorException(ioe.getMessage());
        }

    }

    /**
     * Finds the actual child node that is of the specified element type
     *
     * @param startingNode The node where to find the child
     * @param element      The element type
     *
     * @return a NULL node if childIndex doesn't denote a valid child
     *
     * @exception XMLErrorException Error during XML parsing
     */
    protected Node findNode(Node startingNode, String element)
        throws XMLErrorException {

        Node nodeToFind = null;

        Assert.check(startingNode!=null);

        NodeList children = startingNode.getChildNodes();
        int nbChildren = children.getLength();

        for (int i = 0; i < nbChildren; ++i) {
            Node node = children.item(i);

            if (node.getNodeType() == Node.ELEMENT_NODE) {

                String nodeName = node.getNodeName();

                if (nodeName.equals(element)) {
                    nodeToFind = node;
                    break;
                }

            }

        }

        return nodeToFind;

    }

    /**
     * Returns the actual child node (ELEMENT TYPE) of the specified node.
     *
     * @param node       The parent of the child to get
     * @param childIndex The index of the child to get
     *
     * @return a NULL node if childIndex doesn't denote a valid child
     *
     * @exception XMLErrorException Error during XML parsing
     */
    protected Node getActualChildNode(Node node, int childIndex)
        throws XMLErrorException {

        Node actualChild = null;
        int actualChildIndex = 0;

        Assert.check(node!=null);

        NodeList children = node.getChildNodes();
        int nbChildren = children.getLength();

        for (int i = 0; i < nbChildren; ++i) {

            Node child = children.item(i);

            if (child.getNodeType() == Node.ELEMENT_NODE) {

                if (childIndex != actualChildIndex) {
                    ++actualChildIndex;
                } else {
                   actualChild = child;
                   break;
                }

            }

        }

        return actualChild;

    }

}

File: tags/V8_0s01/SystemMngt/platformsystemlibrary/c++/src/CdmwInitConfiguration.cpp
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI. It is derivative work based on PERCO Copyright (C) THALES
 * 2000-2003. All rights reserved.
 * 
 * Copyright (C) THALES 2004-2005. All rights reserved
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published
 * by the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with CARDAMOM; see the file COPYING. If not, write to the
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */

#include <iostream>
#include <memory>

#include "platformsystemlibrary/CdmwInitConfiguration.hpp"
#include "Foundation/common/Assert.hpp"

#include <util/PlatformUtils.hpp>
#include <util/XMLString.hpp>
#include <util/XMLUniDefs.hpp>
#include <framework/XMLFormatter.hpp>
#include <util/TranscodingException.hpp>

#include <dom/DOM_DOMException.hpp>

#include <parsers/DOMParser.hpp>
#include <dom/DOM.hpp>


// For the DOMTreeErrorReporter class
#include <util/XercesDefs.hpp>
#include <sax/ErrorHandler.hpp>
#include <sax/SAXParseException.hpp>
#include <dom/DOMString.hpp>

namespace Cdmw
{
namespace PlatformMngt
{

/**
*Implementation notes: [if no pertinent write none ]
*<p>
*Portability issues: [if no pertinent write none ]
*<p>
*Code specific to Xerces XML Parser
*/


std::string to_string(const DOMString &s)
    throw (std::bad_alloc)
{
    char *c_str = s.transcode();

    std::string str(c_str);

    delete[] c_str;
    
    return str;
}


class DOMTreeErrorReporter : public ErrorHandler
{
public:
    // -----------------------------------------------------------------------
    //  Constructors and Destructor
    // -----------------------------------------------------------------------
    DOMTreeErrorReporter()
    {
    }

    ~DOMTreeErrorReporter()
    {
    } 


    // -----------------------------------------------------------------------
    //  Implementation of the error handler interface
    // -----------------------------------------------------------------------
    void warning(const SAXParseException& toCatch) { }

    void error(const SAXParseException& toCatch)
    {
        std::string reason(to_string(toCatch.getMessage()));
        CDMW_THROW1(XMLErrorException, reason);
    }

    void fatalError(const SAXParseException& toCatch)
    {
        error(toCatch);
    }

    void resetErrors() { }
};


/*
* The implementation of CdmwInitConfiguration
*/

CdmwInitConfiguration::CdmwInitConfiguration(bool do_validate)
    throw(ConfigurationException, InternalErrorException)
{
    // Initialize the XML4C2 system
    try
    {
        XMLPlatformUtils::Initialize();
    }
    catch(const XMLException& toCatch)
    {
        std::cerr << "Error during Xerces-c Initialization.\n"
             << "  Exception message:"
             << to_string(toCatch.getMessage()) << std::endl;

        CDMW_THROW(InternalErrorException);
    }

    try
    {
        //
        //  Create our parser, then attach an error handler to the parser.
        //  The parser will call back to methods of the ErrorHandler if it
        //  discovers errors during the course of parsing the XML document.
        //
        std::auto_ptr<DOMParser> pParser(new DOMParser());

        if (do_validate){ // TODO: Chack exceptions to be thrown...
            pParser->setValidationScheme(DOMParser::Val_Auto);
            pParser->setDoNamespaces(false);
        }

        std::auto_ptr<DOMTreeErrorReporter> pErrorReporter(new DOMTreeErrorReporter());
        (pParser.get())->setErrorHandler(pErrorReporter.get());

        m_parser = pParser.release();
        m_errorReporter = pErrorReporter.release();
    }
    catch(const std::bad_alloc &)
    {
        XMLPlatformUtils::Terminate();
        CDMW_THROW(InternalErrorException);
    }
}


CdmwInitConfiguration::~CdmwInitConfiguration()
    throw()
{
    delete m_errorReporter;
    delete m_parser;

    XMLPlatformUtils::Terminate();
}


DOM_Node 
CdmwInitConfiguration::findNode(const DOM_Node& startingNode, 
                                const std::string& element,
                                int occurrence)
    throw(XMLErrorException, std::bad_alloc)
{
    DOM_Node nodeToFind;

    CDMW_ASSERT(!startingNode.isNull());

    try
    {
        DOM_NodeList children = startingNode.getChildNodes();

        size_t nbChildren = children.getLength();
        int index = 1;
        for (size_t i = 0; i < nbChildren; ++i)
        {
            DOM_Node node = children.item(i);

            if (node.getNodeType() == DOM_Node::ELEMENT_NODE)
            {
                std::string nodeName(to_string(node.getNodeName()));
                if (nodeName.compare(element) == 0)
                {
                    if (index == occurrence)
                    {                        
                        nodeToFind = node;
                        break;
                    } else {
                        index++;
                    }
                }
            }
        }
    } catch (std::bad_alloc &) {
        throw;
    } catch (const XMLException& e) {
        std::string reason(to_string(e.getMessage()));
        CDMW_THROW1(XMLErrorException, reason);
    } catch (const DOM_DOMException& e) {
        std::string reason(to_string(e.msg));
        CDMW_THROW1(XMLErrorException, reason);
    }

    return nodeToFind;

}


DOM_Node CdmwInitConfiguration::getActualChildNode(const DOM_Node& node, size_t childIndex)
    throw(XMLErrorException)
{
    DOM_Node actualChild;
    size_t actualChildIndex = 0;

    CDMW_ASSERT(!node.isNull());

    try
    {
        DOM_NodeList children = node.getChildNodes();

        size_t nbChildren = children.getLength();

        for (size_t i = 0; i < nbChildren; ++i)
        {
            DOM_Node child = children.item(i);

            if (child.getNodeType() == DOM_Node::ELEMENT_NODE)
            {
                if (childIndex != actualChildIndex)
                {
                    ++actualChildIndex;
                }
                else
                {
                   actualChild = child;
                   break;
                }
            }
        }
    }
    catch (const XMLException& e)
    {
        std::string reason(to_string(e.getMessage()));
        CDMW_THROW1(XMLErrorException, reason);
    }
    catch (const DOM_DOMException& e)
    {
        std::string reason(to_string(e.msg));
        CDMW_THROW1(XMLErrorException, reason);
    }

    return actualChild;
}


CdmwInitConfiguration* 
CdmwInitConfiguration::ReadConfiguration(const std::string& XMLConfigurationFile,
                                          bool do_validate)
    throw(ConfigurationException, OutOfMemoryException, InternalErrorException)
{
    CdmwInitConfiguration* conf = 0;
    try {
        conf = new CdmwInitConfiguration(do_validate);
        conf->readConfigurationFile(XMLConfigurationFile);
    } catch (const std::bad_alloc &) {
        CDMW_THROW(OutOfMemoryException);
    } catch (const OutOfMemoryException & ) {
        throw;
    } catch (const ConfigurationException & ) {
        throw;
    } catch (const InternalErrorException & ) {
        throw;
    }

    return conf;
}


void CdmwInitConfiguration::readConfigurationFile(const std::string& XMLConfigurationFile)
    throw(XMLErrorException, OutOfMemoryException)
{
    try
    {
        m_parser->parse(XMLConfigurationFile.c_str());

        DOM_Document document = m_parser->getDocument();

        DOM_Element rootElement = document.getDocumentElement();

        m_servicesNode = findNode(rootElement, "services");

        m_lifeCycleServiceNode = findNode(m_servicesNode, "lifecycle");

        DOM_Node fault_tolerance = findNode(m_servicesNode, "fault-tolerance");
        if (!fault_tolerance.isNull())
            m_DataStoreServiceNode = findNode(fault_tolerance, "datastores");

    } catch (const std::bad_alloc & ) {
        CDMW_THROW(OutOfMemoryException);
    } catch (const XMLException& e) {
        std::string reason(to_string(e.getMessage()));
        CDMW_THROW1(XMLErrorException, reason);
    } catch (const DOM_DOMException& e) {
        std::string reason(to_string(e.msg));
        CDMW_THROW1(XMLErrorException, reason);
    }

}


std::string 
CdmwInitConfiguration::getServiceAttributeValue(const std::string& service, 
                                                 const std::string& attribute)
    throw (BadParameterException, XMLErrorException, OutOfMemoryException)
{
    bool serviceFound = false;
    bool attributeFound = false;

    std::string attributeValue;

    CDMW_ASSERT(!m_servicesNode.isNull());

    try
    {
        // search the specified service among the children nodes
        DOM_Node serviceNode = findNode(m_servicesNode, service);

        if (!serviceNode.isNull())
        {
            serviceFound = true;

            // search the specified attribute
            DOM_NamedNodeMap attributes = serviceNode.getAttributes();
            DOM_Node attribName = attributes.getNamedItem(attribute.c_str());

            if (!attribName.isNull())
            {
                // attribute found
                attributeFound = true;

                attributeValue = to_string(attribName.getNodeValue());
            }
        }
    } catch (const std::bad_alloc & ) {
        CDMW_THROW(OutOfMemoryException);
    } catch (const OutOfMemoryException & ) {
        throw;
    } catch (const XMLErrorException & ) {
        throw;
    } catch (const XMLException& e) {
        std::string reason(to_string(e.getMessage()));
        CDMW_THROW1(XMLErrorException, reason);
    } catch (const DOM_DOMException& e) {
        std::string reason(to_string(e.msg));
        CDMW_THROW1(XMLErrorException, reason);
    }

    if (!serviceFound)
        CDMW_THROW2(BadParameterException, service, " service not found");

    if (!attributeFound)
        CDMW_THROW2(BadParameterException, attribute, " attribute not found");

    return attributeValue;

}


std::string 
CdmwInitConfiguration::getFrameworkAttributeValue(const std::string& framework,
                                                   size_t frameworkInstance, 
//                                                 const std::string& property, 
                                                   const std::string& attribute)
    throw (BadParameterException, XMLErrorException, OutOfMemoryException)
{

    bool frameworkFound = false;
    bool frameworkInstanceFound = false;
//    bool propertyFound = false;
    bool attributeFound = false;

    std::string attributeValue;

    CDMW_ASSERT(!m_lifeCycleServiceNode.isNull());

    try
    {
        // search the specified service among the children nodes
        DOM_Node frameworkNode = findNode(m_lifeCycleServiceNode, framework);

        if (!frameworkNode.isNull())
        {
            frameworkFound = true;

            // search the specified framework instance
            DOM_Node instance = getActualChildNode(frameworkNode, frameworkInstance);

            if (!instance.isNull())
            {
                frameworkInstanceFound = true;

/*
                // search the specified property
                DOM_Node propertyNode = findNode(instance, property);

                if (!propertyNode.isNull())
                {
                    propertyFound = true;
*/                    

                    // search the specified attribute
                    //DOM_NamedNodeMap attributes = propertyNode.getAttributes();
                    DOM_NamedNodeMap attributes = instance.getAttributes();
                    DOM_Node attribName = attributes.getNamedItem(attribute.c_str());

                    if (!attribName.isNull())
                    {
                        // attribute found
                        attributeFound = true;

                        attributeValue = to_string(attribName.getNodeValue());
                    }
//                }
            }
        }
    } catch (const std::bad_alloc &) {
        CDMW_THROW(OutOfMemoryException);
    } catch (OutOfMemoryException &) {
        throw;
    } catch (XMLErrorException & ) {
        throw;
    } catch (const XMLException& e) {
        std::string reason(to_string(e.getMessage()));
        CDMW_THROW1(XMLErrorException, reason);
    } catch (const DOM_DOMException& e) {
        std::string reason(to_string(e.msg));
        CDMW_THROW1(XMLErrorException, reason);
    }

    if (!frameworkFound)
        CDMW_THROW2(BadParameterException, framework, " framework not found");

    if (!frameworkInstanceFound)
        CDMW_THROW2(BadParameterException, "frameworkInstance" , " invalid index");

/*
    if (!propertyFound)
        CDMW_THROW2(BadParameterException, property, " property not found");
*/

    if (!attributeFound)
        CDMW_THROW2(BadParameterException, attribute, " attribute not found");

    return attributeValue;
}


std::string 
CdmwInitConfiguration::getFrameworkMultithreadAttribute(const std::string& framework,
                                                         size_t frameworkInstance, 
                                                         std::string& threadingPolicy, 
                                                         std::string& threadingAttribute)
    throw (BadParameterException, XMLErrorException, OutOfMemoryException)
{

    bool frameworkFound = false;
    bool frameworkInstanceFound = false;
    bool multithreadFound = false;
    bool policyFound = false;

    std::string attributeValue;

    CDMW_ASSERT(!m_lifeCycleServiceNode.isNull());

    try {
        // search the specified service among the children nodes
        DOM_Node frameworkNode;
        frameworkNode = findNode(m_lifeCycleServiceNode, framework);

        if (!frameworkNode.isNull())
        {
            frameworkFound = true;

            // search the specified factory instance
            DOM_Node instance = getActualChildNode(frameworkNode, frameworkInstance);

            if (!instance.isNull())
            {
                frameworkInstanceFound = true;

                DOM_Node multithreadNode = findNode(instance, "multithread");

                if (!multithreadNode.isNull())
                {
                    multithreadFound = true;

                    DOM_Node policyNode = getActualChildNode(multithreadNode, 0);
                    if (!policyNode.isNull())
                    {
                        policyFound = true;
                        threadingPolicy = to_string(policyNode.getNodeName());

                        // search the unique attribute
                        DOM_NamedNodeMap attributes = policyNode.getAttributes();
                        DOM_Node attribute = attributes.item(0);

                        if (!attribute.isNull())
                        {
                            threadingAttribute = to_string(attribute.getNodeName());
                            attributeValue = to_string(attribute.getNodeValue());
                        }
                        else
                        {
                            // the threading attribute is irrelevant
                            threadingAttribute = "";
                            attributeValue = "";
                        }
                    }
                }
            }
        }
    } catch (const std::bad_alloc & ) {
        CDMW_THROW(OutOfMemoryException);
    } catch (const OutOfMemoryException & ) {
        throw;
    } catch (const XMLErrorException & ) {
        throw;
    } catch (const XMLException& e) {
        std::string reason(to_string(e.getMessage()));
        CDMW_THROW1(XMLErrorException, reason);
    } catch (const DOM_DOMException& e) {
        std::string reason(to_string(e.msg));
        CDMW_THROW1(XMLErrorException, reason);
    }

    if (!frameworkFound)
        CDMW_THROW2(BadParameterException, framework, " framework type not found");

    if (!frameworkInstanceFound)
        CDMW_THROW2(BadParameterException, "frameworkInstance" , " invalid index");

    if (!multithreadFound)
        CDMW_THROW2(BadParameterException, framework , " not multithreaded factory type");

    if (!policyFound)
        CDMW_THROW2(BadParameterException, "Threading policy" , " policy not found");

    return attributeValue;
}




std::string 
CdmwInitConfiguration::getDataStoreAttributeValue(const std::string& datastoreName,
                                                  int datastoreInstance,
                                                  const std::string& attribute)
    throw (BadParameterException, XMLErrorException, OutOfMemoryException)
{
    
    bool datastoreFound = false;
    bool attributeFound = false;

    std::string attributeValue;
    CDMW_ASSERT(!m_DataStoreServiceNode.isNull());

    try
    {
        // search the specified service among the children nodes
        DOM_Node datastoreNode = findNode(m_DataStoreServiceNode, datastoreName, datastoreInstance);
        if (!datastoreNode.isNull())
        {
            datastoreFound = true;

            DOM_NamedNodeMap attributes = datastoreNode.getAttributes();            
            DOM_Node attribName = attributes.getNamedItem(attribute.c_str());

            if (!attribName.isNull())
            {
               // attribute found
                attributeFound = true;
                
                attributeValue = to_string(attribName.getNodeValue());
            } else {
                DOM_Node parentNode = datastoreNode.getParentNode();
                DOM_NamedNodeMap parentAttributes = parentNode.getAttributes();    
                DOM_Node parentAttribName = parentAttributes.getNamedItem(attribute.c_str());

                if (!parentAttribName.isNull())
                {
                    // attribute found
                    attributeFound = true;
                    
                    attributeValue = to_string(parentAttribName.getNodeValue());
                }
            }
        }
    } catch (const std::bad_alloc &) {
        CDMW_THROW(OutOfMemoryException);
    } catch (OutOfMemoryException &) {
        throw;
    } catch (XMLErrorException & ) {
        throw;
    } catch (const XMLException& e) {
        std::string reason(to_string(e.getMessage()));
        CDMW_THROW1(XMLErrorException, reason);
    } catch (const DOM_DOMException& e) {
        std::string reason(to_string(e.msg));
        CDMW_THROW1(XMLErrorException, reason);
    }

    if (!datastoreFound)
        CDMW_THROW2(BadParameterException, datastoreName, " datastore not found");

    if (!attributeFound)
        CDMW_THROW2(BadParameterException, attribute, " attribute not found");

    return attributeValue;
}


std::string 
CdmwInitConfiguration::getDataStoresAttributeValue
(const std::string& attribute)
    throw (BadParameterException, XMLErrorException, OutOfMemoryException)
{
    std::string attributeValue;
    CDMW_ASSERT(!m_DataStoreServiceNode.isNull());

    try
    {
        // search the specified service among the children nodes
        DOM_NamedNodeMap attributes = m_DataStoreServiceNode.getAttributes();
        DOM_Node attribName = attributes.getNamedItem(attribute.c_str());

        if (attribName.isNull())
        {
            CDMW_THROW2(BadParameterException, attribute,
                        " attribute not found");
        }

        attributeValue = to_string(attribName.getNodeValue());
    } catch (const std::bad_alloc &) {
        CDMW_THROW(OutOfMemoryException);
    } catch (OutOfMemoryException &) {
        throw;
    } catch (XMLErrorException & ) {
        throw;
    } catch (const XMLException& e) {
        std::string reason(to_string(e.getMessage()));
        CDMW_THROW1(XMLErrorException, reason);
    } catch (const DOM_DOMException& e) {
        std::string reason(to_string(e.msg));
        CDMW_THROW1(XMLErrorException, reason);
    }

    return attributeValue;    
}


}; // End namespace PlatformMngt
}; // End namespace Cdmw

